### PR TITLE
Rework `Term` visitor to include patterns.

### PR DESCRIPTION
**Pull request by wsx-ucb**
Yak shaving.

### CODE REVIEW

File: base/src/main/java/org/aya/core/def/CtorDef.java [hunk 1]
```diff
@@ -45,6 +46,16 @@ public CtorDef(
     return ref;
   }
 
+  public @NotNull CtorDef update(@NotNull ImmutableSeq<Arg<Pat>> pats, @NotNull ImmutableSeq<Term.Param> selfTele, @NotNull Term result, @NotNull Partial.Split<Term> clauses) {
+    return pats.sameElements(this.pats, true) && selfTele.sameElements(this.selfTele, true) && result == this.result && clauses == this.clauses
+      ? this : new CtorDef(dataRef, ref, pats, ownerTele, selfTele, clauses, result, coerce);
```
Comment (ice1000) [inline base/src/main/java/org/aya/core/def/CtorDef.java:51]: You should never `new CtorDef`! In the constructor of this class, we mutate `ref.core` so the old one will never be referred to

**Review by ice1000 — CHANGES_REQUESTED** [review]

File: base/src/main/java/org/aya/core/def/CtorDef.java (hunk 1 above)
Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/def/CtorDef.java:51]: Will this cause any problem?
I think the updated `CtorDef` is still well-formed, right?
(In the sense that for `updated = old.update(...)`, `updated.ref.core == updated`.)

Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/def/CtorDef.java:51]: Oh I see it now, the reference of all other `DefVar` will now point to the updated definition, which is undesirable.
The same applies to `FnDef` etc, right?

Comment (wsx-ucb) [reply base/src/main/java/org/aya/core/def/CtorDef.java]: Updated the way we visit `GenericDef` to only allow `Consumer`, and no transformations.

**Review by ice1000 — APPROVED** [review]

**Comment by ice1000** [comment]
bors merge

**Pull request merged**